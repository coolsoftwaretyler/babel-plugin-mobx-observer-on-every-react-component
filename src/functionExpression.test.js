import { expect, test, describe} from "bun:test";
import { transform } from "@babel/core";
import pluginSyntaxJsx from "@babel/plugin-syntax-jsx";
import pluginTransReactJsx from "@babel/plugin-transform-react-jsx";
import { autoObserverPlugin } from "./index.mjs";

const runTransform = (input) => {
  return transform(input, {
    plugins: [pluginSyntaxJsx, pluginTransReactJsx, autoObserverPlugin],
  });
};

describe("Mobx Observer Babel Plugin", () => {
  describe("function expression", () => {
    describe("when it is not a React component", () => {
      test("it does nothing", () => {
        const source = `const MyFunctionExpression = function() { return 0 }`;
        const out = runTransform(source);

        expect(out.code).toMatchSnapshot();
      });
    });
    describe("when the component is already wrapped in observer", () => {
      test("it does nothing", () => {
        const source = `observer(function MyComponentExpression() { return <div>Hello World</div>; })`;
        const out = runTransform(source);

        expect(out.code).toMatchSnapshot();
      });
      describe("when the function is anonymous", () => {
        test("it does nothing", () => {
          const source = `const MyAnonymousObserver = observer(function() { return <div>Hello World</div>; })`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe("when the function is an arrow function", () => {
        test("it does nothing", () => {
          const source = `const MyArrowObserver = observer(() => <div>Hello World</div>)`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe("when the function is named", () => {
        test("it does nothing", () => {
          const source = `const MyNamedObserver = observer(function MyNamedFunction() { return <div>Hello World</div>; })`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
    });
    describe("when the component is not wrapped in observer", () => {
      describe("when the function is anonymous", () => {
        test("it wraps the component in observer", () => {
          const source = `const MyAnonymousObserver = function() { return <div>Hello World</div>; }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });

      describe("when the function is named", () => {
        test("it wraps the component in observer", () => {
          const source = `const MyNamedObserver = function MyNamedFunction() { return <div>Hello World</div>; }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
    });
  });
});
