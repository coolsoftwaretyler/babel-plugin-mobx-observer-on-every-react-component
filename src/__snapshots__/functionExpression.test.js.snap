// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobx Observer Babel Plugin function expression when it is not a React component it does nothing 1`] = `
"const MyFunctionExpression = function () {
  return 0;
};"
`;

exports[`Mobx Observer Babel Plugin function expression when the component is already wrapped in observer when the function is anonymous it does nothing 1`] = `
"import { observer } from "mobx-react";
const MyAnonymousObserver = observer(function () {
  return /*#__PURE__*/React.createElement("div", null, "Hello World");
});"
`;

exports[`Mobx Observer Babel Plugin function expression when the component is already wrapped in observer when the function is an arrow function it does nothing 1`] = `
"import { observer } from "mobx-react";
const MyArrowObserver = observer(() => /*#__PURE__*/React.createElement("div", null, "Hello World"));"
`;

exports[`Mobx Observer Babel Plugin function expression when the component is already wrapped in observer when the function is named it does nothing 1`] = `
"import { observer } from "mobx-react";
const MyNamedObserver = observer(function MyNamedFunction() {
  return /*#__PURE__*/React.createElement("div", null, "Hello World");
});"
`;

exports[`Mobx Observer Babel Plugin function expression when the component is already wrapped in observer it does nothing 1`] = `
"import { observer } from "mobx-react";
observer(function MyComponentExpression() {
  return /*#__PURE__*/React.createElement("div", null, "Hello World");
});"
`;

exports[`Mobx Observer Babel Plugin function expression when the component is not wrapped in observer when the function is anonymous it wraps the component in observer 1`] = `
"import { observer } from "mobx-react";
const MyAnonymousObserver = observer(function () {
  return /*#__PURE__*/React.createElement("div", null, "Hello World");
});"
`;

exports[`Mobx Observer Babel Plugin function expression when the component is not wrapped in observer when the function is named it wraps the component in observer 1`] = `
"import { observer } from "mobx-react";
const MyNamedObserver = observer(function MyNamedFunction() {
  return /*#__PURE__*/React.createElement("div", null, "Hello World");
});"
`;
